### PR TITLE
Allow 'expired system timers' to be cancelled before execution.

### LINKED ISSUE #19387: Canceling timers does not work reliably in the LayerImplSelect implementation
#### Problem
When we process timers, we grab a batch out of the list before we start calling them, and now the timers in that batch cannot be canceled.  This can lead to use-after-free.

 #### Proposed Solution
Either stop doing the batch thing and process timers one by one (maybe with a cap on how far down the list we go), or make the batch a member so that canceling can remove from the batch too.  Discussion on Slack tended toward the former solution.

We should also consider making sure that the Select and FreeRTOS implementations are the same, so our CI tells us something about the FreeRTOS case.

@kpschoedel

### LINKED ISSUE #22160: DMalloc detection of free space overwrite during TestRead
#### Problem
 
 Two instances for this at least:
 
 https://github.com/project-chip/connectedhomeip/runs/8005304886?check_suite_focus=true 
 https://github.com/project-chip/connectedhomeip/runs/7984389882?check_suite_focus=true 
 
 Log says:
 
 ```
 ...
 [1661296765.557679][75818:75818] CHIP:EM: Retransmitting MessageCounter:128264913 on exchange 47182i Send Cnt 1
[1661296765.561911][75818:75818] CHIP:IN: Sending encrypted msg 0x7fc57a2058d0 with MessageCounter:128264913 to 0x0000000011223344 (2) at monotonic time: 000000000019EB6C msec
 [1661296765.570471][75818:75818] CHIP:DMG: Unblock report hold after min 57311 seconds
debug-malloc library: dumping program, fatal error
   Error: free space has been overwritten (err 67)
Aborted (core dumped)
INF Test 1/1: [FAIL] TestRead
```

I am unclear where `57311` seconds come from (seems like an abnormally large seconds count) however timestamps do not seem to point at a large thread block either. Unfortunately TestRead in general is very slow to run and with a lot of logs, making it hard to debug.

### PR DESCRIPTION
**Pull request by andy31415**
#### Problem
SystemTimers cannot be reliably cancelled for socket impl.

Specifically to avoid starvation, system timers keeps a list of 'active timers' and that is not cancelable.

#### Change overview
- Add unit test to verify that we can cancel in-progress timers
- Fix timer cancellation (have a member list instead of stack variable for 'in process' list
- Fix unit test for LoopbackTransport (self-cancelling timers on message send scheduling seemed to interfere with tests).

Fixes #22160
Fixes #19387 

#### Testing
Unit tests. CI will also validate.

### CODE REVIEW

**Comment by andy31415** [comment]
1.0 Acceptance: high value fix (fixes test flakyness and potential use after free)

File: src/transport/raw/tests/NetworkTestHelpers.h [hunk 1]
```diff
@@ -68,9 +68,16 @@ class LoopbackTransportDelegate
     // configurable allowed number of messages (mNumMessagesToAllowBeforeDropping)
     virtual void OnMessageDropped() {}
 };
-
 class LoopbackTransport : public Transport::Base
 {
+private:
+    // Use unique pointers for work callbacks, so that one callback does not cancel another.
```
Comment (bzbarsky-apple) [inline src/transport/raw/tests/NetworkTestHelpers.h:74]: This is a workaround for the real issue, which is that ScheduleWork on the select impl is broken.  We should be fixing that, not working around it.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/transport/raw/tests/NetworkTestHelpers.h (hunk 1 above)
Comment (andy31415) [reply src/transport/raw/tests/NetworkTestHelpers.h:74]: Yes... did not try to yak shave into that one...

**Comment by andy31415** [comment]
Closing as duplicate with #22375 and that one fixes more (fixes ScheduleWork too)

**Pull request closed without merging**